Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
From h-shimamoto@ct.jp.nec.com Thu May 15 09:57:50 2008
Date: Thu, 17 Apr 2008 16:57:20 +0200
From: Hiroshi Shimamoto <h-shimamoto@ct.jp.nec.com>
To: Peter Zijlstra <peterz@infradead.org>
Cc: Steven Rostedt <rostedt@goodmis.org>,
     linux-rt-users <linux-rt-users@vger.kernel.org>,
     Ingo Molnar <mingo@elte.hu>, Thomas Gleixner <tglx@linutronix.de>,
     LKML <linux-kernel@vger.kernel.org>
Subject: [PATCH -rt] avoid deadlock related with PG_nonewrefs and swap_lock
Resent-Date: Thu, 17 Apr 2008 14:57:28 +0000 (UTC)
Resent-From: Peter Zijlstra <peterz@infradead.org>
Resent-To: Steven Rostedt <rostedt@goodmis.org>

Hi Peter,

I've updated the patch. Could you please review it?

I'm also thinking that it can be in the mainline because it makes
the lock period shorter, correct?

---
 mm/swapfile.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

--- a/mm/swapfile.c
+++ b/mm/swapfile.c
@@ -373,6 +373,7 @@ int remove_exclusive_swap_page(struct pa
 	/* Is the only swap cache user the cache itself? */
 	retval = 0;
 	if (p->swap_map[swp_offset(entry)] == 1) {
+		spin_unlock(&swap_lock);
 		/* Recheck the page count with the swapcache lock held.. */
 		lock_page_ref_irq(page);
 		if ((page_count(page) == 2) && !PageWriteback(page)) {
@@ -381,8 +382,8 @@ int remove_exclusive_swap_page(struct pa
 			retval = 1;
 		}
 		unlock_page_ref_irq(page);
-	}
-	spin_unlock(&swap_lock);
+	} else
+		spin_unlock(&swap_lock);
 
 	if (retval) {
 		swap_free(entry);
